File: 35.search-insert-position.cpp
/*
 * @lc app=leetcode id=35 lang=cpp
 *
 * [35] Search Insert Position
 */

#include <assert.h>
#include <bits/stdc++.h>

using namespace std;

// @lc code=start
class Solution {
 public:
  int searchInsert(vector<int>& nums, int target) {
    int len = nums.size();
    for (int i = 0; i < len; i++) {
        if (nums[i] >= target) {
            return i;
        }
    }
    return len;
  }
};
// @lc code=end

int main() {
  Solution s;
  vector<int> vec;
  vec = {1, 3, 5, 6};
  assert(2 == s.searchInsert(vec, 5));
  assert(1 == s.searchInsert(vec, 2));
  assert(4 == s.searchInsert(vec, 7));
  assert(0 == s.searchInsert(vec, 0));
  vec = {1};
  assert(0 == s.searchInsert(vec, 0));
  vec = {};
  assert(0 == s.searchInsert(vec, 1));
  vec = {-15, 0, 13};
  assert(1 == s.searchInsert(vec, -13));
}

File: 35.search-insert-position.rs
/*
 * @lc app=leetcode id=35 lang=rust
 *
 * [35] Search Insert Position
 */

struct Solution;

// @lc code=start
impl Solution {
    pub fn search_insert(nums: Vec<i32>, target: i32) -> i32 {
        let mut start = 0;
        let mut end = nums.len() - 1;
        while end > start {
            if nums[start] == target {
                return start as i32;
            } else if nums[end] == target {
                return end as i32;
            } else if nums[start] > target {
                return start as i32;
            } else if nums[end] < target {
                return (end + 1) as i32;
            }
            let mid = (start + end) / 2;
            if nums[mid] < target {
                start = mid + 1;
            } else {
                end = mid - 1;
            }
        }
        if nums[start] < target {
            start += 1;
        }
        start as i32
    }
}
// @lc code=end

#[test]
fn test1() {
    assert_eq!(Solution::search_insert(vec!(1, 3, 5, 6), 5), 2);
}

#[test]
fn test2() {
    assert_eq!(Solution::search_insert(vec!(1, 3, 5, 6), 2), 1);
}

#[test]
fn test3() {
    assert_eq!(Solution::search_insert(vec!(1, 3, 5, 6), 7), 4);
}

#[test]
fn test4() {
    assert_eq!(Solution::search_insert(vec!(1, 3, 5, 6), 0), 0);
}

#[test]
fn test5() {
    assert_eq!(Solution::search_insert(vec!(1), 0), 0);
}
